### PR TITLE
subnets associated with route table need to be removed before route t…

### PR DESCRIPTION
**Pull request by mschurenko**
When trying to delete a route table it fails as subnets are still associated with it.

If I have this in route-test.yml

``` yaml
- ec2_vpc_route_table:
        state: absent
        vpc_id: "{{ vpc_id }}"
        subnets: "{{ subnets }}"
        routes: "{{ routes }}"
        resource_tags:
          Name: test
```

And then run:

``` bash
$ ansible-playbook route-test.yml -vvv
PLAY [test deleting a route with ec2_vpc_route_table] **************************

TASK [ec2_vpc_route_table subnets={{ subnets }} state=absent resource_tags={u'Name': u'test'} routes={{ routes }} vpc_id={{ vpc_id }} region={{ region }}] ***
task path: /Users/mschurenko/git/mschurenko/aws_accounts/route-test.yml:13
ESTABLISH LOCAL CONNECTION FOR USER: mschurenko
localhost EXEC (umask 22 && mkdir -p "`echo $HOME/.ansible/tmp/ansible-tmp-1447808165.89-163262947445826`" && echo "`echo $HOME/.ansible/tmp/ansible-tmp-1447808165.89-163262947445826`")
localhost PUT /var/folders/fl/zr66x4cd3xz_l08tsdmhj6080000gq/T/tmppnHfTX TO /Users/mschurenko/.ansible/tmp/ansible-tmp-1447808165.89-163262947445826/ec2_vpc_route_table
localhost EXEC LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /Users/mschurenko/.ansible/tmp/ansible-tmp-1447808165.89-163262947445826/ec2_vpc_route_table; rm -rf "/Users/mschurenko/.ansible/tmp/ansible-tmp-1447808165.89-163262947445826/" > /dev/null 2>&1
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_args": {"region": "us-west-2", "resource_tags": {"Name": "test"}, "routes": [{"dest": "0.0.0.0/0", "gateway_id": "igw-7fb12b1a"}], "state": "absent", "subnets": "subnet-019ed676", "vpc_id": "vpc-0394b066"}, "module_name": "ec2_vpc_route_table"}, "msg": "The routeTable 'rtb-0e0e396b' has dependencies and cannot be deleted."}

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1
```

This PR runs disassociate_route_table to remove all the existing associated subnets first.

### CODE REVIEW

**Comment by gregdek** [comment]
@erydo @wimnat Please review this PR to make sure it adheres to the following guidelines:

http://docs.ansible.com/developing_modules.html#module-checklist

If it passes these guidelines, and if you believe it’s a good PR otherwise, please add a comment with "shipit" in the text, and we will flag it for inclusion.

If it fails these guidelines, please add a comment with "needs_revision" in the text, along with the needed changes, and we will flag it for revision.

**Comment by wimnat** [comment]
Could you add a force parameter so this can be specified? I.e. if force is true then do as this commit does otherwise fail.  Default to false.

**Comment by mschurenko** [comment]
Sorry @wimnat for not responding to this sooner. 

Why would we want a force option for this? In what scenario would you not want the subnet associated with a route table not to be disassociated first? You simply cannot delete a route table that has associated subnets and it is unlikely that you would have a route table that has no subnet associations as it would serve no purpose. This has no destructive effect on the subnets themselves and is the same behaviour when one deletes a route table via the AWS console.

**Comment by wimnat** [comment]
@mschurenko Think of it this way, if you created a playbook to create your route table you would probably follow this up with some subnet creation.  If i was then to reverse this i'd destroy my subnets then my route table.

Now if i didn't destroy the subnets then tried to destroy the route table at the moment i get an error which is good as it means i can write my code to destroy the subnets first - i have clear infrastructure as code telling me what is going on.  Some [IMO, lazy ;)] people may not care about this and just want to nuke the table and subnets.  Well in that case, this 'force' option would be for them.  Maybe call it 'destroy_subnets'.

needs_revision

**Comment by mschurenko** [comment]
I'm sorry but I really don't understand your point here. Are you concerned that the subnets will be destroyed along with the route table? If so then there is no need for concern as all it does is disassociates the subnet from the route table first. This does not destroy the subnets themselves.

**Comment by mschurenko** [comment]
Maybe this will help make more sense 

http://docs.aws.amazon.com/AWSEC2/latest/CommandLineReference/ApiReference-cmd-DisassociateRouteTable.html

**Comment by gregdek** [comment]
Thanks @mschurenko for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@mschurenko A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

File: cloud/amazon/ec2_vpc_route_table.py [hunk 1]
```diff
@@ -377,6 +378,11 @@ def ensure_subnet_associations(vpc_conn, vpc_id, route_table, subnets,
 
     return {'changed': changed}
 
+def remove_subnet_associations(vpc_conn, vpc_id, route_table, check_mode):
+    for a_id in [a.id for a in route_table.associations]:
+        changed = True
+        vpc_conn.disassociate_route_table(a_id, dry_run=check_mode)
```
Comment (wimnat) [inline cloud/amazon/ec2_vpc_route_table.py:384]: You should wrap this in a try

File: cloud/amazon/ec2_vpc_route_table.py [hunk 2]
```diff
@@ -377,6 +378,11 @@ def ensure_subnet_associations(vpc_conn, vpc_id, route_table, subnets,
 
     return {'changed': changed}
 
+def remove_subnet_associations(vpc_conn, vpc_id, route_table, check_mode):
+    for a_id in [a.id for a in route_table.associations]:
+        changed = True
```
Comment (wimnat) [inline cloud/amazon/ec2_vpc_route_table.py:383]: Changed should be after the boto call

**Comment by gregdek** [comment]
@mschurenko Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@mschurenko A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@mschurenko Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@mschurenko A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@mschurenko Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@mschurenko A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@mschurenko Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
Closing due to contributor inactivity.

**Pull request closed without merging**